File: apps/web/lib/board.ts
import { createEngine, type RawDeparture } from "@signaller/routing-adapter";
import { normaliseCrs, operatorFromRouteName } from "@signaller/shared";
import { darwinStationMessage } from "@signaller/db";
import { sql } from "drizzle-orm";
import { enrichBoardWithDarwin } from "./darwin-board";
import { enrichBoardWithFormation } from "./darwin-formation";
import { enrichBoardWithPosition } from "./board-position";
import { orderByExpectedDeparture } from "./board-order";
import { type Disruption, fetchStationDisruptions } from "./disruptions";
import { track, type Timer } from "./timing";
import { fetchLdbwsBoard, ldbwsConfigured } from "./ldbws";
import { getRtppmByOperatorName } from "./rtppm";
import { stationName } from "./stations";
import { getDb } from "./db";
import { vstpDeparturesForBoard } from "./vstp-board";
import { ukHhmm } from "./uk-time";

/** Where a board's data came from — drives the UI provenance badge. */
export type BoardSource = "ldbws" | "darwin" | "timetable";

/** One carriage in a train formation. */
export interface Coach {
  number: string;
  first: boolean;
  /** Raw class string from the feed, e.g. "First" / "Standard". */
  coachClass?: string;
  /** Loading percentage (0-100) when the operator reports it. */
  loading?: number;
}

/** A single row on a live departure board. */
export interface BoardDeparture {
  tripId?: string;
  /** Darwin run id, once the live overlay has matched this trip. */
  rid?: string;
  originName?: string;
  originCrs?: string;
  destinationName: string;
  destinationCrs?: string;
  operator?: string;
  /** ISO instant. Never a bare "HH:MM" — see `live`. */
  scheduled: string;
  /**
   * Live estimated departure when Darwin has one, as an ISO INSTANT.
   *
   * Every consumer does `new Date(live)`. darwin-board.ts used to assign the
   * raw "HH:MM" string here, which produced Invalid Date in the board's
   * formatter and a silently blank countdown; keep this an instant.
   */
  live?: string;
  platform?: string;
  platformChanged: boolean;
  /** Human status: on time, Exp. HH:MM, Cancelled, Delayed. */
  status: "on-time" | "delayed" | "cancelled" | "scheduled";
  delayMinutes?: number;
  /** Human-readable cause, e.g. "delayed by a late running train in front". */
  reason?: string;
  /** Number of coaches, when the operator reports a formation. */
  coachCount?: number;
  /** Per-coach detail (number, class, live loading) when available. */
  coaches?: Coach[];
  /** True once real-time data has been applied to this row. */
  hasLive: boolean;
  /** Live "where is it right now" from Network Rail (correlated by rid). */
  position?: BoardPosition;
  /** Operator's rolling last-hour punctuality (RTPPM), 0-100, when known. */
  operatorPunctuality?: number;
  /**
   * True when live running has moved this row out of its booked position.
   * The board orders by expected departure, so a delayed train drops down the
   * list — the flag lets the UI say that rather than appear to shuffle.
   */
  movedFromSchedule?: boolean;
}

/** Live running position for a board row, from the Network Rail overlay. */
export interface BoardPosition {
  /** Human status: where the train physically is at the moment. */
  label: string;
  /** Seconds late (+) / early (−) at the last report. */
  latenessMinutes?: number;
  /** How long ago Network Rail last reported it, in seconds. */
  reportedAgoSeconds?: number;
  /** True while the train hasn't left its origin yet. */
  approaching: boolean;
  /**
   * True when the last report is old enough that the position may no longer be
   * where the train actually is. The row is still shown — a quiet feed is not
   * the same as a train that vanished — but the UI must say the age out loud
   * rather than present a stale location as current.
   */
  stale?: boolean;
}

export interface BoardResult {
  crs: string;
  stationName: string;
  generatedAt: string;
  /** True when the board carries real-time data (LDBWS, or Darwin overlay). */
  live: boolean;
  source: BoardSource;
  /** When filtered, the name of the "calling at" station. */
  filterName?: string;
  /** Station disruption notices (NRCC messages from LDBWS). */
  messages: string[];
  /** Structured active disruptions from the Disruptions API. */
  disruptions: Disruption[];
  departures: BoardDeparture[];
  arrivals: BoardDeparture[];
}

export type BoardOutcome =
  | { ok: true; board: BoardResult }
  | { ok: false; reason: "engine-offline" | "bad-request" | "unknown-station" };

/** NRCC station messages for the MOTIS-fallback board path (LDBWS already carries its own). */
async function fetchStationMessages(crs: string): Promise<string[]> {
  const db = getDb();
  try {
    const rows = await db
      .select({ html: darwinStationMessage.html })
      .from(darwinStationMessage)
      .where(sql`${darwinStationMessage.crsList} @> ARRAY[${crs}]::text[]`);
    return rows.map((r) => r.html);
  } catch {
    return [];
  }
}

/**
 * Loosely-matchable key for an operator display name — strips case, spacing
 * and punctuation so "South Western Railway" (LDBWS) and the Disruptions
 * API's own tocName line up even if one has a stray "&"/hyphen the other
 * doesn't. Exported for the commute network-log filter (commute-network-log.ts),
 * which does the same RTPPM-style fuzzy join against a different feed.
 */
export function normaliseOperatorName(name: string): string {
  return name.toLowerCase().replace(/[^a-z0-9]/g, "");
}

/** Attach each row's operator's rolling punctuality figure, where matched. */
async function withPunctuality(departures: BoardDeparture[]): Promise<BoardDeparture[]> {
  if (departures.length === 0) return departures;
  const rtppm = await getRtppmByOperatorName().catch(() => new Map());
  if (rtppm.size === 0) return departures;
  return departures.map((d) => {
    if (!d.operator) return d;
    const match = rtppm.get(normaliseOperatorName(d.operator));
    if (!match?.rollingPpm) return d;
    return { ...d, operatorPunctuality: Math.round(match.rollingPpm) };
  });
}

function scheduledRow(raw: RawDeparture): BoardDeparture {
  return {
    tripId: raw.tripId,
    destinationName: raw.destinationName ?? raw.headsign ?? "",
    destinationCrs: raw.destinationStopId,
    operator: operatorFromRouteName(raw.routeName),
    scheduled: raw.scheduled,
    live: raw.live,
    // The engine's "platform" is the .MSN station-level stop_desc (identical
    // for every train here), not a real platform number — real platforms come
    // from Darwin. Don't show the misleading placeholder.
    platform: undefined,
    platformChanged: false,
    status: raw.cancelled ? "cancelled" : "scheduled",
    hasLive: false,
  };
}

function scheduledArrivalRow(raw: RawDeparture): BoardDeparture {
  return {
    tripId: raw.tripId,
    originName: raw.originName ?? raw.headsign ?? "",
    originCrs: raw.originStopId,
    destinationName: raw.destinationName ?? "",
    destinationCrs: raw.destinationStopId,
    operator: operatorFromRouteName(raw.routeName),
    scheduled: raw.scheduled,
    live: raw.live,
    platform: undefined,
    platformChanged: false,
    status: raw.cancelled ? "cancelled" : "scheduled",
    hasLive: false,
  };
}

/**
 * Order rows by when they will actually leave, tagging any the live picture
 * has displaced. Applied to both source paths so the two boards behave alike.
 */
function inExpectedOrder(departures: BoardDeparture[]): BoardDeparture[] {
  return orderByExpectedDeparture(departures).map(({ departure, movedFromSchedule }) =>
    movedFromSchedule ? { ...departure, movedFromSchedule } : departure,
  );
}

function inExpectedArrivalOrder(arrivals: BoardDeparture[]): BoardDeparture[] {
  return orderByExpectedDeparture(arrivals).map(({ departure }) => departure);
}

function departureKey(d: BoardDeparture): string {
  const minute = ukHhmm(d.scheduled) ?? d.scheduled;
  return `${minute}|${d.destinationCrs ?? ""}|${d.destinationName.toLowerCase()}`;
}

function mergeVstpDepartures(
  scheduled: BoardDeparture[],
  vstp: BoardDeparture[],
): BoardDeparture[] {
  if (vstp.length === 0) return scheduled;
  const seen = new Set(scheduled.map(departureKey));
  const merged = [...scheduled];
  for (const row of vstp) {
    const key = departureKey(row);
    if (seen.has(key)) continue;
    seen.add(key);
    merged.push(row);
  }
  return merged;
}

export async function getBoard(
  crsInput: string,
  when?: string,
  limit = 20,
  callingAt?: string,
  /**
   * Optional. When passed, each stage of the pipeline is timed and reported in
   * the response's Server-Timing header. Callers that don't care omit it and
   * pay nothing.
   */
  timer?: Timer,
): Promise<BoardOutcome> {
  let crs: string;
  try {
    crs = normaliseCrs(crsInput);
  } catch {
    return { ok: false, reason: "bad-request" };
  }

  const name = await stationName(crs);
  // stationName echoes the CRS back when it doesn't know the station.
  if (name === crs) return { ok: false, reason: "unknown-station" };

  // Normalise an optional "calling at" filter station.
  let filterCrs: string | undefined;
  if (callingAt) {
    try {
      filterCrs = normaliseCrs(callingAt);
    } catch {
      filterCrs = undefined;
    }
  }

  // Active disruptions for this station, fetched alongside the board (never
  // blocks it — returns [] on any failure).
  const disruptionsPromise = fetchStationDisruptions(crs).catch(() => []);

  // Primary source: LDBWS live board (real platforms, live estimates, messages).
  // "when" is only meaningful to the timetable engine — LDBWS is always "now".
  if (ldbwsConfigured() && !when) {
    const ldbws = await track(timer, "ldbws", fetchLdbwsBoard(crs, limit, filterCrs));
    // With a filter, an empty result is authoritative ("nothing calls there") —
    // don't fall through to the unfiltered MOTIS board. Without a filter, an
    // empty LDBWS result means fall back.
    if (ldbws && (ldbws.departures.length > 0 || ldbws.arrivals.length > 0 || filterCrs)) {
      for (const d of ldbws.departures) {
        if (!d.destinationName && d.destinationCrs) {
          d.destinationName = await stationName(d.destinationCrs);
        }
      }
      const ldbwsArrivals = filterCrs
        ? ((await track(timer, "ldbws-arrivals", fetchLdbwsBoard(crs, limit, filterCrs, "from")))
            ?.arrivals ?? [])
        : ldbws.arrivals;
      for (const d of ldbwsArrivals) {
        if (!d.originName && d.originCrs) {
          d.originName = await stationName(d.originCrs);
        }
        if (!d.destinationName && d.destinationCrs) {
          d.destinationName = await stationName(d.destinationCrs);
        }
      }
      // Fill coach formation from Darwin where LDBWS didn't provide it, then
      // overlay the live Network Rail running position (where the train is now).
      const withFormation = await track(
        timer,
        "formation",
        enrichBoardWithFormation(crs, ldbws.departures),
      );
      const withPosition = await track(
        timer,
        "position",
        enrichBoardWithPosition(crs, withFormation),
      );
      const departures = inExpectedOrder(
        await track(timer, "punctuality", withPunctuality(withPosition)),
      );
      const arrivals = inExpectedArrivalOrder(await withPunctuality(ldbwsArrivals));
      return {
        ok: true,
        board: {
          crs,
          stationName: ldbws.stationName || name,
          generatedAt: ldbws.generatedAt,
          live: true,
          source: "ldbws",
          filterName: ldbws.filterName,
          messages: ldbws.messages,
          disruptions: await disruptionsPromise,
          departures,
          arrivals,
        },
      };
    }
  }

  // Fallback: MOTIS scheduled board, overlaid with any Darwin DB forecasts.
  //
  // createEngine() is inside the try on purpose: it throws when MOTIS_URL is
  // unset, and that used to escape as a 500. An unconfigured routing engine is
  // the same situation for a caller as an unreachable one, so both report
  // engine-offline and the board degrades instead of erroring.
  let raw: RawDeparture[];
  let rawArrivals: RawDeparture[];
  try {
    const engine = createEngine();
    [raw, rawArrivals] = await track(
      timer,
      "motis",
      Promise.all([
        engine.departures({ stopId: crs, when, limit }),
        engine.arrivals({ stopId: crs, when, limit }),
      ]),
    );
  } catch {
    return { ok: false, reason: "engine-offline" };
  }

  const scheduled = raw.map(scheduledRow);
  const scheduledArrivals = rawArrivals
    .map(scheduledArrivalRow)
    .filter((row) => !filterCrs || row.originCrs === filterCrs);
  const vstp = when
    ? []
    : await track(timer, "vstp", vstpDeparturesForBoard(crs, new Date(), limit).catch(() => []));
  const scheduledWithVstp = mergeVstpDepartures(scheduled, vstp);
  const { departures: darwinDepartures, live } = await track(
    timer,
    "darwin",
    enrichBoardWithDarwin(crs, scheduledWithVstp),
  );

  for (const d of darwinDepartures) {
    if (!d.destinationName && d.destinationCrs) {
      d.destinationName = await stationName(d.destinationCrs);
    }
  }
  for (const d of scheduledArrivals) {
    if (!d.originName && d.originCrs) {
      d.originName = await stationName(d.originCrs);
    }
  }

  // Same formation/position enrichment as the LDBWS path — previously skipped
  // here, leaving fallback boards without coach detail or live NR position.
  const withFormation = await track(
    timer,
    "formation",
    enrichBoardWithFormation(crs, darwinDepartures),
  );
  const withPosition = await track(timer, "position", enrichBoardWithPosition(crs, withFormation));
  const departures = inExpectedOrder(
    await track(timer, "punctuality", withPunctuality(withPosition)),
  );
  const arrivals = inExpectedArrivalOrder(
    await track(timer, "arrival-punctuality", withPunctuality(scheduledArrivals)),
  );

  return {
    ok: true,
    board: {
      crs,
      stationName: name,
      generatedAt: new Date().toISOString(),
      live,
      source: live ? "darwin" : "timetable",
      messages: await fetchStationMessages(crs),
      disruptions: await disruptionsPromise,
      departures,
      arrivals,
    },
  };
}

File: apps/ios/Signaller/Models/Board.swift
import Foundation

/// Mirrors `BoardResult` / `BoardDeparture` in `apps/web/lib/board.ts`.
///
/// The previous client decoded a narrow subset and dropped `messages`,
/// `disruptions`, `arrivals`, `platformChanged`, `reason`, `coaches` and
/// `position` — which is most of what makes the web board useful. They're all
/// here now.

struct BoardResponse: Codable {
    let ok: Bool
    let board: Board?
    let reason: String?
}

struct Board: Codable, Hashable {
    let crs: String
    let stationName: String
    let generatedAt: Date
    /// True when the board carries real-time data (LDBWS, or Darwin overlay).
    let live: Bool
    let source: String
    /// When filtered, the name of the "calling at" station.
    let filterName: String?
    /// NRCC station messages. May contain HTML.
    let messages: [String]
    let disruptions: [Disruption]
    let departures: [Departure]
    let arrivals: [Departure]
}

struct Disruption: Codable, Hashable, Identifiable {
    var id: String { (title ?? "") + (summary ?? "") }
    let title: String?
    let summary: String?
    let url: String?
}

struct Departure: Codable, Identifiable, Hashable {
    /// The API gives no stable row id, so fall back through the identifiers it
    /// does provide.
    var id: String {
        rid ?? tripId ?? "\(scheduled.timeIntervalSince1970)-\(destinationName)"
    }

    let tripId: String?
    /// Darwin run id, once the live overlay has matched this trip.
    let rid: String?
    let originName: String?
    let originCrs: String?
    let destinationName: String
    let destinationCrs: String?
    let operatorName: String?
    let scheduled: Date
    /// Live estimate. An instant, not "HH:MM".
    let live: Date?
    let platform: String?
    let platformChanged: Bool
    let status: JourneyStatus
    let delayMinutes: Int?
    /// Human-readable cause, e.g. "delayed by a late running train in front".
    let reason: String?
    let coachCount: Int?
    let coaches: [Coach]?
    /// True once real-time data has been applied to this row.
    let hasLive: Bool
    let position: BoardPosition?
    /// Operator's rolling last-hour punctuality (RTPPM), 0-100.
    let operatorPunctuality: Double?
    /// True when live running moved this row out of its booked position. The
    /// board sorts by expected departure, so a delayed train drops down the
    /// list — this flag lets the UI say so rather than appear to shuffle.
    let movedFromSchedule: Bool?

    /// The time to show: live when present.
    var effectiveTime: Date { live ?? scheduled }

    /// True when live running moved the time enough to be worth showing both.
    var isRetimed: Bool {
        guard let live else { return false }
        return abs(live.timeIntervalSince(scheduled)) >= 60
    }

    private enum CodingKeys: String, CodingKey {
        case tripId, rid, originName, originCrs, destinationName, destinationCrs
        case scheduled, live, platform, platformChanged, status, delayMinutes
        case reason, coachCount, coaches, hasLive, position, operatorPunctuality
        case movedFromSchedule
        case operatorName = "operator"
    }

    /// A minimal stand-in built from a rid, for opening the service screen from
    /// somewhere that has no board row — a train tapped on the map. Only `rid`
    /// and the destination are known; the screen fetches the rest.
    static func placeholder(rid: String, destination: String) -> Departure {
        Departure(
            tripId: nil,
            rid: rid,
            originName: nil,
            originCrs: nil,
            destinationName: destination,
            destinationCrs: nil,
            operatorName: nil,
            scheduled: Date(),
            live: nil,
            platform: nil,
            platformChanged: false,
            status: .scheduled,
            delayMinutes: nil,
            reason: nil,
            coachCount: nil,
            coaches: nil,
            hasLive: true,
            position: nil,
            operatorPunctuality: nil,
            movedFromSchedule: nil
        )
    }
}

struct Coach: Codable, Hashable, Identifiable {
    /// Stable for the life of the value.
    ///
    /// This was `number ?? UUID().uuidString` — a *computed* property, so an
    /// unnumbered coach minted a fresh id on every access. In a `ForEach` that
    /// destroys and rebuilds the row every render. Not reachable today (the
    /// formation strip renders `ServiceCoach`), but a trap for the next caller.
    let id: String
    let number: String?
    let coachClass: String?
    let toilet: String?
    /// Live loading as a percentage, when the operator reports it.
    let loading: Int?

    private enum CodingKeys: String, CodingKey {
        case number, toilet, loading
        case coachClass = "class"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        number = try container.decodeIfPresent(String.self, forKey: .number)
        coachClass = try container.decodeIfPresent(String.self, forKey: .coachClass)
        toilet = try container.decodeIfPresent(String.self, forKey: .toilet)
        loading = try container.decodeIfPresent(Int.self, forKey: .loading)
        // Assigned once at decode. An unnumbered coach still gets a stable id
        // for the lifetime of the value, which is what `ForEach` needs.
        id = number ?? UUID().uuidString
    }
}

/// Live "where is it right now", from the Network Rail overlay.
struct BoardPosition: Codable, Hashable {
    let label: String
    /// Minutes late (+) / early (−) at the last report.
    let latenessMinutes: Double?
    /// How long ago Network Rail last reported it.
    let reportedAgoSeconds: Double?
    /// True while the train hasn't left its origin yet.
    let approaching: Bool
    /// True when the last report is old enough that the train may have moved
    /// on. The row is still shown — a quiet feed isn't a vanished train — but
    /// the UI must say the age out loud rather than imply this is current.
    let stale: Bool?
}
